core/oa_hash_map.h

```diff
@@ -0,0 +1,571 @@
+	uint32_t rehash_position;
+	uint32_t rehash_amount;
+
+	uint32_t elements;
+
+	/* Methods */
+
+	// returns true if the value already existed, false if it's a new entry
+	bool _raw_set_with_hash(uint32_t p_hash, const TKey &p_key, const TData &p_data) {
+		for (int i = 0; i < table.capacity; i++) {
+
+			int pos = (p_hash + i) % table.capacity;
+
+			int flags_pos = pos / 4;
+			int flags_pos_offset = pos % 4;
+
+			bool is_filled_flag = table.flags[flags_pos] & (1 << (2 * flags_pos_offset));
+			bool is_deleted_flag = table.flags[flags_pos] & (1 << (2 * flags_pos_offset + 1));
+
+			if (is_filled_flag) {
+				if (table.hashes[pos] == p_hash && Comparator::compare(table.keys[pos], p_key)) {
+					table.data[pos] = p_data;
+					return true;
+				}
+				continue;
+			}
+
+			table.keys[pos] = p_key;
+			table.data[pos] = p_data;
+			table.hashes[pos] = p_hash;
+
+			table.flags[flags_pos] |= (1 << (2 * flags_pos_offset));
+			table.flags[flags_pos] &= ~(1 << (2 * flags_pos_offset + 1));
+
+			return false;
+		}
+		return false;
+	}
+
+	void _rehash_progress() {
+		for (int i = 0; i <= rehash_amount && rehash_position < old_table.capacity; rehash_position++) {
+
+			int flags_pos = rehash_position / 4;
+			int flags_pos_offset = rehash_position % 4;
+
+			bool is_filled_flag = (old_table.flags[flags_pos] & (1 << (2 * flags_pos_offset))) > 0;
+			bool is_deleted_flag = (old_table.flags[flags_pos] & (1 << (2 * flags_pos_offset + 1))) > 0;
+
+			if (is_filled_flag) {
+				_raw_set_with_hash(old_table.hashes[rehash_position], old_table.keys[rehash_position], old_table.data[rehash_position]);
+
+				old_table.keys[rehash_position].~TKey();
+				old_table.data[rehash_position].~TData();
+
+				memnew_placement(&old_table.keys[rehash_position], TKey);
+				memnew_placement(&old_table.data[rehash_position], TData);
+
+				old_table.flags[flags_pos] &= ~(1 << (2 * flags_pos_offset));
+				old_table.flags[flags_pos] |= (1 << (2 * flags_pos_offset + 1));
+			}
+		}
+
+		if (rehash_position >= old_table.capacity) {
+
+			// wohooo, we can get rid of the old table.
+			is_rehashing = false;
+
+#ifdef OA_HASH_MAP_INITIAL_LOCAL_STORAGE
+			if (old_table.data == local_data) {
+				// Everything is local, so no cleanup :P
+			} else
+#endif
+			{
+				memdelete_arr(old_table.data);
+				memdelete_arr(old_table.keys);
+				memdelete_arr(old_table.hashes);
+				memdelete_arr(old_table.flags);
+			}
+		}
+	}
+
+public:
+	_FORCE_INLINE_ uint32_t get_capacity() const { return table.capacity; }
+	_FORCE_INLINE_ uint32_t get_num_elements() const { return elements; }
+
+	void set(const TKey &p_key, const TData &p_data) {
+
+		uint32_t hash = Hasher::hash(p_key);
+
+		// We don't progress the rehashing if the table just got resized
+		// to keep the cost of this function low.
+		if (is_rehashing) {
+
+			// rehash progress
+			_rehash_progress();
+		}
+
+		// Table is almost full, resize and start rehashing process.
+		if (elements >= table.capacity * 0.7) {
```

Comment:
Would it make sense to benchmark the table with various fill levels? I suspect that moving this number higher or lower based off of the quality of our hashfunctions may help here.

Reply:
Given that the [Wikipedia article on Open Addressing](https://en.wikipedia.org/wiki/Open_addressing) states "Even with good hash functions, load factors are normally limited to 80%", I think limiting it to 70%-80% is good enough.

Reply:
Yes, some place on the internet (*caugh* Wikipedia *caugh*) said that after 0.7 fill ratio the local probing needs to go ahead so far that some of the cache locality benefits go away compared to chaining.